test/test_autograd.py

```diff
@@ -5721,6 +5776,49 @@ def test_copy_(self, device):
             z = x.to(torch.bfloat16)
             self.assertTrue(z.requires_grad)
 
+    @onlyCUDA
+    def test_simple_reentrant_cross_device(self, device):
+        class ReentrantFunc(Function):
+            _cpu_mode = True
+            @staticmethod
+            def forward(ctx, x):
+                with torch.enable_grad():
+                    ctx.x = Variable(x, requires_grad=True)
+                    ctx.y = Variable(x + 2, requires_grad=True)
+                    ctx.output_var = ctx.x * ctx.y
+                return ctx.output_var.detach()
+
+            @staticmethod
+            def backward(ctx, grad_output):
+                with torch.enable_grad():
+                    new_param = ctx.output_var.detach().requires_grad_()
+                    if ReentrantFunc._cpu_mode:
+                        new_param = new_param.cpu()
+                        (new_param ** 2).sum().backward()
+                    else:
+                        new_param = new_param.cuda()
+                        (new_param ** 2).sum().backward()
+                return grad_output
+
+        # Reentrant starts on GPU thread, finishs on GPU thread
+        x = torch.randn(2, 2, device=device, requires_grad=True)
+        out = ReentrantFunc.apply(x)
+        out.sum().backward()
+
+        # Reentrant starts on CPU thread, finishs on GPU thread
```

Comment:
This one will actually finish on the CPU no?
Because the AccumulateGrad for the leaf you create here `new_param = ctx.output_var.detach().requires_grad_()` will run on the CPU.

Reply:
updated with new tensors instead.